linera-core/src/client.rs

```diff
@@ -1076,78 +1079,111 @@ where
         tracker: u64,
         committees: BTreeMap<Epoch, Committee>,
         max_epoch: Epoch,
-        node: impl LocalValidatorNode,
+        node: impl LocalValidatorNode + Clone,
         node_client: LocalNodeClient<S>,
     ) -> Result<(ValidatorName, u64, Vec<Certificate>), NodeError> {
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_nth(tracker);
         let response = node.handle_chain_info_query(query).await?;
         // Responses are authenticated for accountability.
         response.check(&name)?;
-        let mut certificates = Vec::new();
-        let mut new_tracker = tracker;
-        for entry in response.info.requested_received_log {
-            let query = ChainInfoQuery::new(entry.chain_id)
-                .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
-            let local_response = node_client
-                .handle_chain_info_query(query.clone())
-                .await
-                .map_err(|error| NodeError::LocalNodeQuery {
-                    error: error.to_string(),
-                })?;
-            if !local_response
-                .info
-                .requested_sent_certificate_hashes
-                .is_empty()
-            {
-                new_tracker += 1;
-                continue;
-            }
-
-            let mut response = node.handle_chain_info_query(query).await?;
-            let Some(certificate_hash) = response.info.requested_sent_certificate_hashes.pop()
-            else {
-                break;
-            };
+        let log = response.info.requested_received_log;
+        let mut new_tracker = tracker + log.len() as u64;
+        let mut stream = stream::iter(log.into_iter().zip(tracker..))
+            .map(|(entry, index)| {
+                let node_client = node_client.clone();
+                let node = node.clone();
```

Comment:
Isn't it expensive to clone client/nodes?

Reply:
I don't think so, but good question. It probably comes down to how expensive it is to clone a `tonic::client::Grpc`.